lib/sources.nix

```diff
@@ -26,12 +26,31 @@ rec {
     (type == "symlink" && lib.hasPrefix "result" baseName)
   );
 
-  cleanSource = builtins.filterSource cleanSourceFilter;
+  cleanSource = filterSource cleanSourceFilter;
+
+  # Like `builtins.filterSource`, except it will compose with itself,
+  # allowing you to chain multiple calls together without any
+  # intermediate copies being put in the nix store.
+  filterSource = f: src:
+    let isFiltered = src ? _is-lib-filterSource;
+    in assert isLocalSource src; rec {
```

Comment:
Hmm what does this assertion actually gain us? `builtins.filterSource` itself will throw here. And note that `typeOf src == "path"` is not quite exhaustive, as you can call `builtins.filterSource` on, say `toString ./.` (though I'm not sure if there are actual cases where that matters)

Reply:
Hm. Yea I suppose the only way to exhaustively check that `builtins.filterSource` can take the argument would be to actually try it and catch any thrown error; but that would defeat the point since it would evaluate the intermediate filtered sources.

I was using the assertion to avoid an error message that's difficult to debug, but you're right; I don't think it will work. But the question remains: How do we have `callCabal2nix` determine when it's appropriate to try to filter?

Reply:
`callCabal2nix` can do something like check `(src.type or "") != "derivation"`

Reply:
That's not exhaustive though... I think either you have to do something like check if src is a string and if so if it's in the nix store, or just take an arg to callCabal2nix.

Reply:
@edolstra Would you object to a `hasContext` primop? That would solve this...